File: RideShareApp/RideShareApp/Rider.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace RideShareApp
{
    public class Rider : User
    {
        public string Location { get; set; }
        public void RequestRide(RideType rideType, string pickup, string dropOff)
        {
            Console.WriteLine($"{Name} requested a {rideType} ride from {pickup} to {dropOff}");
        }

        public void RateDriver(Driver driver, double rating)
        {
            driver.Rating = (driver.Rating + rating) / 2;
            Console.WriteLine($"{Name} rated driver {driver.Name} with {rating} stars");
        }

        public void MakePayment(IPaymentMethod paymentMethod, double fare)
        {
            paymentMethod.ProcessPayment(fare, this);
        }
       
    }
}

File: RideSharingApp__Strategy_Pattern/src/Rider.java
public class Rider extends User {
    private IPaymentMethod preferredPaymentMethod;

    public Rider(int id, String name, String location, IPaymentMethod preferredPaymentMethod) {
        super(id, name, location);
        this.preferredPaymentMethod = preferredPaymentMethod;
    }

    public void requestRide(String pickupLocation, String dropOffLocation, RideType rideType) {
        System.out.println(name + " has requested a " + rideType + " ride from " + pickupLocation + " to " + dropOffLocation + ".");
    }

    public void rateDriver(Driver driver, double rating) {
        driver.updateRating(rating);
        System.out.println(name + " rated " + driver.getName() + " with " + rating + " stars.");
    }

    public void makePayment(Trip trip) {
        preferredPaymentMethod.ProcessPayment(trip.getFare());
    }

    // Method to change the payment method dynamically
    public void changePaymentMethod(IPaymentMethod newMethod) {
        this.preferredPaymentMethod = newMethod;
        System.out.println(name + " selected a payment method.");
    }

    @Override
    public void displayInfo() {
        System.out.println("Rider: " + name + ", Location: " + location + ", Rating: " + rating);
    }
}
